Physics/IntensityBuilderXML.hpp

```diff
@@ -19,48 +20,51 @@
 namespace ComPWA {
 class Kinematics;
 
-namespace FunctionTree {
-class OldIntensity;
-class FunctionTreeIntensityWrapper;
-} // namespace FunctionTree
-
 namespace Tools {
 class IntegrationStrategy;
 }
 
 namespace Physics {
-class NamedAmplitude;
-
 namespace HelicityFormalism {
 class HelicityKinematics;
 }
 
+struct IntensityBuilderState {
+  ComPWA::FunctionTree::ParameterList Parameters;
+  ComPWA::FunctionTree::ParameterList Data;
+};
+
+
+// user needs couple of building functions:
+// 1. build intensity + kinematics (for all kinds of things, but mainly data
+// generation and plotting). this needs just a model file and potentially a phsp
+// sample
+// 2. build estimator + fit parameter list (for fitting). this needs a model
+// file, a data sample (and potentially a phsp sample, which depends on the
+// estimator implemtnation)
 class IntensityBuilderXML {
 public:
   IntensityBuilderXML(const std::vector<Event> phspsample = {});
 
-  std::tuple<
-      std::shared_ptr<ComPWA::FunctionTree::FunctionTreeIntensityWrapper>,
-      std::shared_ptr<HelicityFormalism::HelicityKinematics>>
-  createIntensityAndKinematics(const boost::property_tree::ptree &pt) const;
+  std::pair<ComPWA::FunctionTree::FunctionTreeIntensityWrapper,
+            HelicityFormalism::HelicityKinematics>
+  createIntensityAndKinematics(const boost::property_tree::ptree &pt);
 
-  std::shared_ptr<HelicityFormalism::HelicityKinematics>
+  HelicityFormalism::HelicityKinematics
   createHelicityKinematics(std::shared_ptr<PartList> partL,
-                           const boost::property_tree::ptree &pt) const;
-  std::shared_ptr<ComPWA::FunctionTree::FunctionTreeIntensityWrapper>
-  createIntensity(std::shared_ptr<PartList> partL,
-                  std::shared_ptr<Kinematics> kin,
-                  const boost::property_tree::ptree &pt) const;
-
-  std::shared_ptr<ComPWA::FunctionTree::OldIntensity>
-  createOldIntensity(std::shared_ptr<PartList> partL,
-                     std::shared_ptr<Kinematics> kin,
-                     const boost::property_tree::ptree &pt) const;
-
-  std::shared_ptr<NamedAmplitude>
-  createAmplitude(std::shared_ptr<PartList> partL,
-                  std::shared_ptr<Kinematics> kin,
-                  const boost::property_tree::ptree &pt) const;
+                           const boost::property_tree::ptree &pt);
+
+  std::pair<ComPWA::FunctionTree::FunctionTreeEstimatorWrapper,
+            FitParameterList>
+  createMinLogLHFunctionTreeEstimator(
```

Comment:
Shouldn't that function be part of MinLogLH?

Reply:
Yeah, thats the part I'm still working on/thinking about. The problem is that you need to make a copy of the Intensity FunctionTree for the MinLogLH Estimator. Because you need to evaluate once the actual data sample and once the phsp sample. Since only the Intensity Builder has the FunctionTree information (which is currently private, since the building functions return the FunctionTrees are private), the creation of the MinLogLH Estimator has to go inside this class.
I dont really like this either. Maybe we have to make the FunctionTree building function public. Or do you have any ideas?

Reply:
We could just build two Intensities, one for the data sample and one for the phase space sample. I mean if you have one tree which is one evaluated with the data and and afterwards with the phase space sample, than you basically don't cache anything, since the tree is recalculated at each change of the data sample

Reply:
Yeah that's what I'm doing right now already. Its just a matter of code design.

Reply:
I just had an idea. The comment you made above would also be solved this way. What if we adjust the FunctionTree backend this way that on a copy operation an actual copy will be made. Usually the move operation will do just fine. But in case of creating a estimator making a copy would be needed. I think that makes things really transparent and easy. What do you think?

Reply:
I'm not sure if I understand what you mean. Why is a copy mandatory for an estimator?

Reply:
What I meant with that is, that for the construction of the MinLogLH you would need to copy your Intensity once and evaluate that copy with the phsp sample instead.
If we use the approach with the copy operator/constructor, we have the advantage to only make a single implementation for a concrete estimator (ie MinLogLH). That implementation could be used with any Intensity no matter what actually performs the calculation underneath (functiontree etc)

Reply:
We have to consider that the FunctionTree can become very memory consuming (depending on the amplitude model). I think we should try to avoid a "deep" copy. Probably we could delete the copy constructor to ensure that only a move is possible.

Reply:
Well for the construction of the MinLogLH estimator, a copy of the function tree is inevitable. No matter what you do, one copy has to be made. I agree that a "deep" copy operation could lead to misuse. Although I don't see it as a problem because:
- Unless you write bad code, the object would not be copied (but moved or reference)
- Its natural to assume, that a copy is a "deep" copy. This is just the c++ way
For me the pros outweigh the cons here. Which solution do you recommend instead?

Reply:
The full FunctionTree should not reside in memory twice. I think thats very important.

Reply:
Hmm you mean it should be impossible (deleting copy constructor)? If I imagine everything correctly, our code will not generated it twice. However the user might create a copy of it...
If we want to delete the copy constructor then we have to search for another solution.

The large memory consumption is also a problem of the function tree right now, since everything is cached. I think if we implement some wiser caching, the consumption will be much lower, without losing any speed (maybe even faster).

Reply:
The current implementation - as far as I understand it - allows to copy the FunctionTree via the standart copy operator i.e. the shared_ptr's for two copies point to the same objects. Thats fine for me, but one could say it is a bit misleading since - as you said - one would expect that a 'deep' copy is done. 
* I would disfavour the solution in which a 'deep' copy is implemented since it is possible to create the tree several times in memory and I do not see a nessesity to have this functionality.
* Deleting the copy constructor would make clear that the tree should never be copied.
But as I said the current code is fine for me

Reply:
Hmm ok, then I have to come up with something better. The current solution is not very flexible for the creation of the estimators, because right now they have to be part of the `Builder`. Currently I'm thinking about your idea (@weidenka) to normalize every intensity automatically on creation. Then the estimator would only have to evaluate the Intensity and thats it. Also the estimators would not need to know about the underlying implementation of Intensity. Seems perfect! I can't think of any downside of this so far? Can you guys? Maybe this is the best solution for the problem.
I will give this a try next week, and hopefully I can wrap up this pull request then.

Reply:
That sounds good to me

Reply:
> Also the estimators would not need to know about the underlying implementation of Intensity.

Actually this statement is not true. In that way the FunctionTree will not perform caching for the data evaluation. In order to get the full caching feature, a full FunctionTree estimator has to be created. That only works by having the information of the intensity FunctionTree in the first place. This also makes the creation process of the estimators equally difficult (similar problems as the other solutions).

Reply:
So this part is also "finished" now. The outermost intensity is automatically normalized (even if the no NormalizedIntensity tag is used). I still had to give the `FunctionTreeEstimatorWrapper` a public access to the FunctionTree members, which is not good. But I don't see any way around this right now, except what I described above.